### ISSUE TITLE
Automatic Mirror update rate limiting?

### ISSUE THREAD

**Issue by gavin-norman-sociomantic** [sociomantic-tsunami/dhtproto#9]
The current Mirror request (also the legacy Listen request), which forwards every single change on the listened channel, requires us to limit the rate of updates to heavily mirrored channels, as the update rate has an exponential knock-on effect on the DHT (1 updated record results in `n` writes, where `n` is the number of listeners).

It would be nicer for the update rate to be set on the receiving side. Something like:
- Each Mirror request handler maintains a list of keys of records which have been updated. Each key can only be in the list once.
- Every x seconds (specified by the client), the handler iterates over this list, fetches the current values of the records, and sends them to the client. The list is cleared afterwards.

**Comment by gavin-norman-sociomantic** [comment]
A bit more info on the situation in which this feature is required.

The channel for which output limiting (at most X records per second) is used by the writer app has a small number of records. It is perfectly possible for the writer to store a pending update for _every_ record in the channel in memory.

**Comment by gavin-norman-sociomantic** [comment]
It would be pretty easy to adapt the current Mirror code to pause sending of updates to the client for a period.

**Comment by gavin-norman-sociomantic** [comment]
This:
>Each Mirror request handler maintains a list of keys of records which have been updated. Each key can only be in the list once.

...plus this:
>It would be pretty easy to adapt the current Mirror code to pause sending of updates to the client for a period.

...would work flawlessly for channels with less records than the maximum length of the update queue in the Mirror handler. (The queue could never overflow, no matter how long the delay before sending.)

**Comment by gavin-norman-sociomantic** [comment]
A simplistic approach might be:
1. The client specifies the maximum number of updates per second.
2. The request handler in the node maintains a list of keys of records which have been updated. Each key can only be in the list once.
3. Once per second, up to the specified number of keys are popped from the list of updated records and sent to the client.
4. Update queue overflows are handled as usual. (The client is warned and updates are lost.)

The problem is, though, what can we do if the queue overflows? Increasing the allowed rate of updates/second would alleviate the problem, but this would defeat the point of having the feature in the first place. Ditto for reducing the rate at which the writer updates the channel.

**Comment by gavin-norman-sociomantic** [comment]
Another aspect to consider: while requiring the writer to limit the update rate of a channel obviously isn't ideal, is requiring the readers to do the same that much better? As mentioned in the issue description, the reason for wanting to limit the update rate at all is for the sake of overall system / DHT performance. I think this problem is actually a lot trickier than it sounds... :/

**Comment by andrej-mitrovic-sociomantic** [comment]
> requiring the writer to limit the update rate of a channel obviously isn't ideal

There's another benefit to having this on the writer side; making the application more efficient. If a single profile is updated very frequently (think 10 times a second) but the reader applications don't need such high-resolution updates, then the app can do the more optimal thing by updating this profile internally 10 times, but have it only write it once to the DHT (e.g. once a second max).

This is assuming a 1-second delay between internall updates and writes is ok with reader apps. For the channels which the app in question writes even multiple-minute delays tend to be ok.

**Comment by gavin-norman-sociomantic** [comment]
Right. I feel like for now, unless someone has a clever idea of how to make this all happen automagically (without anyone having to explicitly specify update rates), I'll move this out of the neo-alpha-4 milestone. Let's leave it as is (the writer limits the update rate) unless we really think of a better solution.

**Comment by nemanja-boric-sociomantic** [comment]
> There's another benefit to having this on the writer side; making the application more efficient. If a single profile is updated very frequently (think 10 times a second) but the reader applications don't need such high-resolution updates, then the app can do the more optimal thing by updating this profile internally 10 times, but have it only write it once to the DHT (e.g. once a second max).

Yes, but (theoretically speaking) same behavior prevents readers to get the updates as quick as it's possible if they want, since the writer will not write them as they are ready. Of course, if such client exists, it's a question if they would (and should) use DHT to communicate with the writer.

**Comment by gavin-norman-sociomantic** [comment]
Indeed, this is one potential pro of allowing the update rate to be specified by the reader: each reader may need a different rate of updates.

**Comment by gavin-norman-sociomantic** [comment]
Thinking about this again, for channels where the main point is to aggregate multiple changes to the same record into a single mirror update, the following might work:

1. The request handler in the node maintains a list of keys of records which have been updated. Each key can only be in the list once.
2. When the list of updated records passes a certain size (half of its possible maximum?), then it is sent to the client.
3. Optionally, the node might also flush pending updates periodically, to make sure updates don't get "stuck" in the list.

That way, whenever a record is updated, there'll be a period where the node waits for further updates to the same record before forwarding the new value on to clients.

I'm not sure if this covers a real use case that we have, but it sounds like it would be helpful in certain circumstances and would be fairly simple to implement.
